maglev: support setting a weight of a backend in a service spec via new cmdline argument

oblazek opened the pull request:
Support setting a weight of a backend when creating via ServiceSpec. This is a full support for backend weights and finishes up work (https://github.com/cilium/cilium/pull/13943) initially done by fankaixi.li (https://github.com/lrouter) and is targeted to XDP L4LB users. K8s support will be done in a following PR.

This way user can set a different backend weight based on its performance or put backend out of service (weight 0) for e.g. maintenance which is important when running in lb-only mode (XDP L4LB).

Usually a node-port (e.g. maglev) service using cli is created by using this command:
```
# cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port
```
this by default sets weight 1 to each backend.
```
# cilium bpf lb maglev get 1                                                                                                                       
SVC ID   LOOKUP TABLE                                                                                                                                                                                               
1        [3 3 3 3 3 4 4 4 4 4 4 3 3 3 3 3 3 3 4 4 4 4 4 3 3 3 3 3 3 3 3 3 4 4 4 4 3 3 3 3 3 3 3 3 3 3 4 4 3 3 3 3 3 3 3 3 3 3 3 4 3 3 3 3 3 3 3 3 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 3 3 4 4 4 4 4 4 4 4 4 4 3 3 3 4 
4 4 4 4 4 4 4 4 4 3 3 3 3 4 4 4 4 4 4 4 4 3 3 3 3 3 3 4 4 4 4 4 4 3 3 3 3 3 3 3 4 4 4 4 4 3 3 3 3 3 3 3 3 3 4 4 4 4 3 3 3 3 3 3 3 3 3 3 4 4 3 3 3 3 3 3 3 3 3 3 3 4 3 3 3 3 3 3 3 3 3 3 3 4 4 4 4 4 4 4 4 4 4 4 4 4 
4 3 3 4 4 4 4 4 4 4 4 4 4 3 3 3 4 4 4 4 4 4 4 4 4 4 3 3 3 3 4 4 4 4 4 4 4 4 3 3 3 3 3 3 4 4 4 4 4 4 3 3 3 3 3 3 3 4 4 4 4 4 4 3 3 3 3 3 3 3 3 4 4 4 4 3 3 3 3 3 3 3 3 3 4 4 4 3 3 3 3 3 3 3 3 3 3 3 4 4 3 3 3 3 3 3 
3 3 3 3 3 3 3 4 4 4 4 4 4 4 4 4 4 4 3 4 4 4 4 4 4 4 4 4 4 4 4 3 3 4 4 4 4 4 4 4 4 4 4 3
```
(output truncated).

with this feature it is possible to define weight of each backend:
```
# cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047,10.246.1.34:7047" --k8s-node-port --backend-weights 20,1,4
```

```
# cilium bpf lb maglev get 1
SVC ID   LOOKUP TABLE
[1]/v4   [1 3 1 1 1 3 1 1 1 1 1 1 1 1 1 1 3 1 1 1 1 1 1 1 1 1 1 1 3 1 1 3 1 1 1 1 2 1 1 3 1 1 1 1 1 1 3 1 1 1 3 1 1 1 1 1 1 1 1 1 1 3 1 2 1 1 1 1 1 1 1 1 3 1 1 1 3 1 1 1 1 1 1 1 3 1 1 3 1 1 2 1 1 1 1 3 1 1 1 1 1
1 3 1 1 1 3 1 1 1 3 1 1 1 1 1 1 3 1 1 1 3 3 1 1 1 1 1 1 1 1 1 3 3 1 1 1 1 1 1 3 1 1 1 2 1 1 3 1 1 1 3 1 1 1 3 1 1 2 1 1 1 3 1 1 1 3 3 1 1 1 2 1 1 1 1 1 3 3 1 1 1 1 1 1 1 1 1 1 3 1 1 3 1 1 1 3 1 2 1 3 1 1 1 1 2 1
3 1 1 1 3 1 1 1 1 1 1 1 1 1 1 3 3 1 2 1 1 1 1 1 1 1 1 3 1 1 3 1 1 1 1 1 1 1 3 1 1 1 1 1 1 3 1 1 1 3 1 1 1 1 3 1 1 1 1 1 3 1 2 1 3 1 1 1 1 1 1 1 3 1 1 3 1 1 1 1 1 3 1 3 1 1 1 1 1 2 1 1 1 1 3 1 1 1 3 1 1 1 1 1 1 3
```
(output truncated)

```
# cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047,10.246.1.34:7047" --k8s-node-port --backend-weights 200,2,0
```

```
# cilium bpf lb maglev get 1                                                                                                                                        
SVC ID   LOOKUP TABLE                                                                                                                                                                                               
[1]/v4   [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 
1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 
```
(output truncated)

Implementation reuses a backend state API (https://github.com/cilium/cilium/pull/18814) when weight 0 is set.

Service list currently doesn't print out the backend weight as there is no way to actually recover it from (backend weight is not stored in bpf map). But this is fine, `cilium bpf lb maglev list/get` can be used instead.

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>
Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

Commit:
api: refactor backend weight scheme to use uint16

Signed-off-by: fankaixi.li <fankaixi.li@bytedance.com>
Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
service: Support setting a weight of a backend in a service spec

Support setting a weight of a backend when creating via ServiceSpec.

This adds a support for backend weights and changes the maglev
algorithm to honor the weights by changing the frequency of backend
turns. This way user can set a different weight for each backend
(e.g. if some of them are less performant than others).

It is also possible to set weight 0 which reuses existing
backend states API and puts backend out for maintenance.
This is important when running in lb-only maglev mode (XDP L4LB).

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
test: add unit tests for maglev weighted backends

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

brb commented on Documentation/cmdref/cilium_service_update.md:
````diff
@@ -11,6 +11,7 @@ cilium service update [flags]
 ### Options
 
 ```
+      --backend-weights uints       Backend weights (0 means maintenance) (default [])
````
Could you mention that for now it either 0 or 1.

oblazek replied:
yeah, will do that, thanks

brb reviewed (changes requested):
The core logic looks good to me. Perhaps you could extend the CI L4LB to set the weight to 0 and to check that requests are failing.

brb reviewed (changes requested):
Thanks. I did another round of reviews:
- Perhaps you could squash the commits into two - `refactor: change UpsertServiceParams Backends field` and `service: Support setting a weight of a backend in a service spec`.
- We should make it more explicit that the backend weight feature is only for the Maglev mode. Could you update the PR title and also the feature flag description to reflect that?

jibi commented on api/v1/models/backend_address.go:
```diff
@@ -29,6 +29,9 @@ type BackendAddress struct {
 	// Layer 4 port number
 	Port uint16 `json:"port,omitempty"`
+
+	// Backend weight
```
Should this be updated to reflect the fact that only `0` and `1` values are currently supported?

oblazek replied:
:+1: SG

jibi commented on cilium/cmd/service_update.go:
```diff
@@ -137,15 +139,26 @@ func updateService(cmd *cobra.Command, args []string) {
 
+	if len(backendWeights) == 0 {
+		backendWeights = make([]uint, len(backends))
+		fmt.Printf("Backend weights not set. Setting all weights to 1\n")
```
Nit: probably not needed as that's the expected default

oblazek replied:
makes sense, thanks

jibi commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -186,6 +186,8 @@ type ID uint32
 type Backend struct {
 	// ID of the backend
 	ID BackendID
+	// Weight of backend
```
(ditto)

jibi commented on pkg/service/service.go:
```diff
@@ -363,8 +363,7 @@ func (s *Service) UpsertService(params *lb.SVC) (bool, lb.ID, error) {
 		// only use node-local backends for external scope. We implement this by
 		// filtering out all backend IPs which are not a local endpoint.
-		// We also filter out backends that have weight equal to 0.
-		if filterBackends && len(b.NodeName) > 0 && b.NodeName != nodeTypes.GetName() || b.Weight == 0 {
```
(probably committed by mistake?)

oblazek replied:
yeah, that was before I realized I need to move the logic closer to maglev table population

jibi commented on pkg/service/service.go:
```diff
@@ -785,19 +785,18 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 		// will happen once the agent receives a delete event for the service
 		// endpoint as it'll be part of obsoleteBackendIDs/obsoleteSVCBackendIDs
 		// list passed to this function.
-		if !b.Terminating && b.Weight > 0 {
```
Why is this getting removed?

oblazek replied:
the same as the comment above: I realized then I need to move the logic closer to maglev table population

jibi commented on Documentation/cmdref/cilium_service_update.md:
````diff
@@ -11,6 +11,7 @@ cilium service update [flags]
 ### Options
 
 ```
+      --backend-weights uints       Backend weights (0 means maintenance) (default [])
````
This commit can probably be squashed with the one that introduced the new CLI option

oblazek committed:
feat: add XDP L4LB e2e test for backend-weights

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

brb commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -152,23 +152,32 @@ func (lbmap *LBBPFMap) UpsertService(p *UpsertServiceParams) error {
+	if len(backendNames) > 0 {
+		sort.Strings(backendNames)
+		table := maglev.GetLookupTable(backendNames, lbmap.maglevTableSize)
+		// In case there are no backends we should delete maglevBackendIDsBuffer
```
Mind elaborating on this? The buffer is global, i.e., it's used by all services. I guess the comment might be misleading.

oblazek replied:
you are so right, sorry that comment is not supposed to be here :(

brb reviewed (approved):
Thanks!

Also, cc @aditighag MBOI

jibi reviewed (changes requested):
@oblazek I think something is not quite right with these changes :thinking: 

I tested this PR starting from a couple of dev VMs based on master:

```
➜  cilium git:(master) NETNEXT=1 NFS=1 K8S=1 NWORKERS=1 VM_CPUS=8 VM_MEMORY=6144 ./contrib/vagrant/start.sh
```

created a few test services (just used the manifest in the test directory):

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ k apply -f test/k8sT/manifests/echo-svc.yaml
```

and enabled maglev in `/etc/sysconfig/cilium`:

```
CILIUM_OPTS="--debug --k8s-require-ipv4-pod-cidr --enable-bandwidth-manager --kube-proxy-replacement=strict --enable-remote-node-identity --tunnel=disabled --ipv4-native-routing-cidr=10.0.0.0/8 --k8s-kubeconfig-path /var/lib/cilium/cilium.kubeconfig --identity-allocation-mode=crd --enable-k8s-event-handover=false --enable-ipv6=false --bpf-lb-algorithm maglev --bpf-lb-maglev-table-size 251"
```

and the LB table seems correctly populated:

```
vagrant@k8s1:~$ sudo cilium bpf lb maglev list
SVC ID   LOOKUP TABLE
14       [4 6 4 6 4 4 4 4 4 6 4 6 4 4 4 4 4 4 6 6 6 6 6 4 6 4 6 4 4 6 6 6 6 4 6 4 6 4 4 6 6 6 6 4 6 4 6 4 4 4 6 6 6 4 6 4 6 4 4 4 6 6 6 4 6 4 6 4 4 4 4 6 6 4 6 4 6 4 4 4 4 6 6 4 6 4 6 4 4 4 4 4 6 4 6 4 6 4 4 4 4 4 6 6 6 6 6 6 4 6 4 6 4 6 6 6 6 6 4 6 4 6 4 6 6 6 6 6 4 6 4 6 4 6 6 6 6 6 4 6 4 4 4 4 6 6 4 6 4 6 4 4 4 4 6 6 4 6 4 6 4 4 4 4 4 6 4 6 4 6 4 4 4 4 4 6 4 6 4 6 4 4 4 4 4 4 6 6 6 6 6 4 6 4 6 4 6 6 6 6 6 4 6 4 6 4 4 6 6 6 6 4 6 4 6 4 4 6 6 6 6 4 6 4 6 4 4 4 6 6 6 4 6 4 6 4 4 4 6 6 6 4 6 4 6 4 4 4 6 6 6]
[..]
```

But after switching to this branch:

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ git checkout oblazek/ob-backend-weight-zero
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ make build
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ sudo systemctl restart cilium
```

the table seems empty:
```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ sudo cilium bpf lb maglev list
SVC ID   LOOKUP TABLE
vagrant@k8s1:~/go/src/github.com/cilium/cilium$
```

also, I see this on the logs:
```
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 9: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 6: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 7: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 8: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 11: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 12: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:58 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 13: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
Feb 21 09:40:59 k8s1 cilium-agent[161715]: level=error msg="Error while inserting service in LB map" error="Unable to delete maglev lookup table 14: delete: key does not exist" k8sNamespace=default k8sSvcName=echo subsys=k8s-watcher
```

not entirely sure what's going on at the moment (and I probably won't have the time to dig further into this in the upcoming days :disappointed:)

aditighag commented on api/v1/openapi.yaml:
```diff
@@ -2466,6 +2466,10 @@ definitions:
         description: Optional name of the node on which this backend runs
         type: string
+      weight:
+        description: Backend weight {0, 1}
```
There is no validation done to ensure that the weight is set to 0 or 1. This also brings up a question about what the full range of weight should be. If it's 0 to 1, then uint16 isn't the right type.

oblazek replied:
I guess you are right, let me add the validation. About the uint16, the initial PR https://github.com/cilium/cilium/pull/13943 uses uint32 which I think is too much. We often use weight up to 1000 (in IPVS setup) and I plan to finish up the full weight support in the future.

oblazek replied:
ah, one more sec.. the validation changed uint to pointer..

oblazek replied:
should be gtg

nathanjsweet reviewed (approved):
LGTM for K8s related things.

aditighag reviewed (changes requested):
We discussed this during the sig-datapath meeting on 03/17. Per the discussion, following changes will be made for the PR - 
- Set Maintenance and Active for backend weights 0 and 1, respectively. The states will be restored after agent restart.  
- Drop changes in `UpsertMaglevLookupTable`. 
- Full range of weights will be supported in future. 

@oblazek Feel free to mark the PR as draft until you've pushed the new changes.

brb commented on bpf/lib/common.h:
```diff
@@ -827,6 +828,8 @@ struct lb4_backend {
 	__be16 port;		/* L4 port filter */
 	__u8 proto;		/* L4 protocol, currently not used (set to 0) */
 	__u8 flags;
+    __u16 weight;   /* Service endpoint weight */
```
Why do you need to add the weight field to the backend struct?

oblazek replied:
I don't think there is another way to store that. If I restart the agent how will I know the weight of each backend? I am not using k8s services for this so there is no way to sync that.

oblazek replied:
other than storing it in backend bpf map

brb replied:
I think we should store in BPF maps only those values which are used by the
datapath. Therefore, the weights should be stored somewhere else.

For the standalone LB case, we assume that it is a user's responsibility to
persist LB services and their configuration.

oblazek replied:
aah, ok.. that simplifies things a lot

oblazek replied:
I thought about it in a way that I cannot afford to loose a service configuration, but you are right, I dont have to

aditighag commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -284,6 +284,7 @@ func (*LBBPFMap) DeleteService(svc loadbalancer.L3n4AddrID, backendCount int, us
 
 // AddBackend adds a backend into a BPF map. ipv6 indicates if the backend needs
 // to be added in the v4 or v6 backend map.
+// AddBackend adds a backend into a BPF map.
```
Nit: Remove the duplicate comment.

aditighag commented on pkg/service/service.go:
```diff
@@ -1423,6 +1424,7 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
 						backend.L3n4Addr, err)
 				}
 				backends[i].ID = id
+				backends[i].Weight = backend.Weight
```
Where do you update the backend weight when user updates it for an existing backend via the command line?

oblazek replied:
yes exactly, I do something like: 
```
cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port --backend-weights 5,100
```

aditighag replied:
I meant that you are only updating backend weights only for new backends. What happens when a user updates weight for an existing backend?
Example - 

```
cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port --backend-weights 5,100

cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port --backend-weights 10,200
```

oblazek replied:
no problem, that works fine

oblazek replied:
will post "bpf lb maglev list" output later

oblazek replied:
```
# cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port --backend-weights 22,100
```
```
level=debug msg="Upserting service" backends="[{ 0 22  {10.246.3.34 {NONE 7047} 0} 0 false} { 0 100  {10.246.39.33 {NONE 7047} 0} 0 false}]" l7LBFrontendPorts="[]" l7LBProxyPort=0 loadBalancerSourceRanges="[]" serviceIP="{10.248.11.13 {TCP 7047} 0}" serviceName= serviceNamespace= sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcHealthCheckNodePort=0 svcTrafficPolicy=Cluster svcType=NodePort
```
```
# cilium bpf lb maglev get 1                                                                                                                                                    
SVC ID   LOOKUP TABLE                                                                                                                                                                                              
[1]/v4   [2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 1 1 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 2 2 1
2 2 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 1 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2
2 1 2 2 2 2 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 1 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 1 2 2 2 2 2 2
2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 1 2 2 2 2 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 1 2 1
2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 1 2 2
2 2 2 1 2 1 2 2 2 2 1 2 2 2 1 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 2 2 1 2
2 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 1 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 2 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 1 2 2 2 2 2 2 1 2 1 2 2 2 2 1 2 2 2 2 2 2 1 2 2 2 2 2 1 2 2
```
and right afterwards I can update the same set of backends...

```
# cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port --backend-weights 90,8
```

```
level=debug msg="Upserting service" backends="[{ 0 90  {10.246.3.34 {NONE 7047} 0} 0 false} { 0 8  {10.246.39.33 {NONE 7047} 0} 0 false}]" l7LBFrontendPorts="[]" l7LBProxyPort=0 loadBalancerSourceRanges="[]" serviceIP="{10.248.11.13 {TCP 7047} 0}" serviceName= serviceNamespace= sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcHealthCheckNodePort=0 svcTrafficPolicy=Cluster svcType=NodePort
```

```
# cilium bpf lb maglev get 1                                                                                                                                                    
SVC ID   LOOKUP TABLE                                                                                                                                                                                              
[1]/v4   [1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1
1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1
1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1
1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1
1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1
1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 
```

aditighag replied:
When I originally added this comment, I didn't remember seeing updates of the backend weights for existing backends. I see it in the latest revisions.

oblazek commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -641,10 +638,10 @@ func getBackend(backend loadbalancer.Backend, ipv6 bool) (Backend, error) {
 	if ipv6 {
 		lbBackend, err = NewBackend6V2(backend.ID, backend.IP, backend.Port, u8proto.ANY,
-			backend.State)
+			backend.State, backend.Weight)
```
this is not used, removing...

oblazek commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -641,10 +638,10 @@ func getBackend(backend loadbalancer.Backend, ipv6 bool) (Backend, error) {
 	} else {
 		lbBackend, err = NewBackend4V2(backend.ID, backend.IP, backend.Port, u8proto.ANY,
-			backend.State)
+			backend.State, backend.Weight)
```
this is not used, removing...

aditighag commented on pkg/service/service.go:
```diff
@@ -1466,43 +1473,38 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
+			// to set the backend state to the saved state.
+			if (backends[i].State == lb.BackendStateTerminating &&
+				b.State != lb.BackendStateTerminating) ||
+				b.Weight == 0 || b.Weight != backends[i].Weight {
```
What the weight related checks are added to the code that updates persisted state? Is this a leftover from previous iterations where you were persisting backend weights in BPF maps?

oblazek replied:
this is similar to the comment below.. 
Let's say I have existing backend. The agent has been restarted. I now want to set a weight of a backend to 0 for a specific service, but as I am not modifying backend state across all services the state would not be changed if the `||
b.Weight == 0` was not there and the latter wouldn't work as the backend weight is not recovered from anywhere.

aditighag replied:
Is there a better way to handle these updates without overloading this code path with weights related checks? If we are not persisting backend weights, why do we want to persist the backend states derived from weight related API updates? We would want to have a consistent behavior across k8s and standalone LB modes. 
What do you think of updating the weights once all the services and their backends have been restored from the datapath - https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L869?

oblazek replied:
I mean I could just remove this part if you think it's no good. We will use `cilium service update --backends "10.246.3.34:7047" --states active` anyway (after or before agent restart) to put backend out/back for/from maintenance.

oblazek replied:
> What do you think of updating the weights once all the services and their backends have been restored from the datapath - https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L869?

I don't believe this makes sense. At this point, in func `RestoreServices()` we have zero knowledge about backend weights. So we are really just waiting for user to update a specific service and/or their backend weights.

aditighag replied:
> We will use `cilium service update --backends "10.246.3.34:7047" --states active` anyway (after or before agent restart) to put backend out/back for/from maintenance.

If you are going to anyway use the `--states` API, then in what cases is this code path needed to persist backend states derived from weights related updates?

> So we are really just waiting for user to update a specific service and/or their backend weights.

What do you mean by this? Aren't users expected to set backend weights after agent restart anyway because the weights are not persisted?

oblazek replied:
> If you are going to anyway use the --states API, then in what cases is this code path needed to persist backend states derived from weights related updates?

Well only for the case, when backend is put e.g. into maintenance, cilium-agent restarted, and user would want to put this backend into active state **only** for specific service (not for all). 

> What do you mean by this? Aren't users expected to set backend weights after agent restart anyway because the weights are not persisted?

No they are not expected to set backend weights after agent restart as they do not have to. If they haven't changed the maglev lookup table size the lut stays intact even after agent restart and the backend weights are only needed to modify the maglev lut.

aditighag replied:
>No they are not expected to set backend weights after agent restart as they do not have to.

If you don't update weights, the second time the agent restarts, the agent will set backend weights to 0, and recalculate maglev table accordingly?

aditighag replied:
We need to have a common logic that works for both k8s and standalone modes. Here are the code changes I propose  - 

- Add a flag to track whether backend weights are explicitly set. Set this flag whenever weights are set/updated for backends in the API handler. 
- Read this flag to determine the case where weights are updated, hence the saved backend weight and state need to be updated. 
- We can get rid of the `RestoredFromDatapath` flag, but we don't have to do that in your PR.

Here is the diff (doesn't include the API handler changes, but should be trivial):

<details><summary>proposed diff</summary>

diff --git a/pkg/loadbalancer/loadbalancer.go b/pkg/loadbalancer/loadbalancer.go
index dfdc77b1ae..c4275b7d01 100644
--- a/pkg/loadbalancer/loadbalancer.go
+++ b/pkg/loadbalancer/loadbalancer.go
@@ -310,7 +310,8 @@ type Backend struct {
        // ID of the backend
        ID BackendID
        // Weight of backend
-       Weight uint16
+       Weight        uint16
+       WeightUpdated bool
        // Node hosting this backend. This is used to determine backends local to
        // a node.
        NodeName string
diff --git a/pkg/service/service.go b/pkg/service/service.go
index dac44636cd..cb3848138d 100644
--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -1483,7 +1483,7 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
                        svc.backendByHash[hash] = &backends[i]
                } else {
                        backends[i].ID = b.ID
-                       b.RestoredFromDatapath = false
+                       updateSavedBeState := false
                        // Backend state can either be updated via kubernetes events,
                        // or service API. If the state update is coming via kubernetes events,
                        // then we need to update the internal state. Currently, the only state
@@ -1491,9 +1491,15 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
                        // weight has changed or is set to zero (maintenance). All other state
                        // updates happen via the API (UpdateBackendsState) in which case we need
                        // to set the backend state to the saved state.
-                       if (backends[i].State == lb.BackendStateTerminating &&
-                               b.State != lb.BackendStateTerminating) ||
-                               b.Weight == 0 || b.Weight != backends[i].Weight {
+                       switch {
+                       case backends[i].State == lb.BackendStateTerminating &&
+                               b.State != lb.BackendStateTerminating:
+                               updateSavedBeState = true
+                       case backends[i].WeightUpdated:
+                               updateSavedBeState = true
+                               b.Weight = backends[i].Weight
+                       }
+                       if updateSavedBeState {
                                b.State = backends[i].State
                                // Update the persisted backend state in BPF maps.
                                if err := s.lbmap.UpdateBackendWithState(backends[i]); err != nil {
@@ -1504,7 +1510,6 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
                                // Set the backend state to the saved state.
                                backends[i].State = b.State
                        }
-                       b.Weight = backends[i].Weight
                }
        }

</details>

aditighag commented on pkg/service/service.go:
```diff
@@ -1466,43 +1473,38 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
 		} else {
 			backends[i].ID = b.ID
-			// Update backend state.
-			if b.RestoredFromDatapath {
-				backends[i].State = b.State
-				// Toggle the flag as the backend is now restored.
-				b.RestoredFromDatapath = false
```
Can you explain why you removed this code path? Was this intentional?

oblazek replied:
Yeah this was intentional, so let's say I:
1) Add a service with some number of backends - all backend weight 1. All is good (all state active).
2) I restart the cilium-agent.
3) Now I want to put one of the backends out for maintenance. I run:
```
cilium service update --id 1 --frontend "10.248.11.13:7047" --backends "10.246.3.34:7047,10.246.39.33:7047" --k8s-node-port --backend-weights 1,0
```
But if I keep this code as is, the backend at this stage has flag `RestoredFromDatapath` set to true, which means the backend state is replaced here by the previous state (the one before restart, which was active) and this means I have to run this (the `cilium service update`...) twice to actually change the backend state.

*) with the changes I did:
I am able to run this only once and there is no problem AFAICS.

aditighag replied:
This breaks the existing logic (runtime CI test should fail) where we restore backend states from BPF maps. If you notice the code path , once the backends are restored after agent restart, the `RestoredFromDatapath` flag is toggled. You would need to re-apply backend weights once all the service backends are restored - https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L869.

oblazek replied:
Why do you think so? The flag `RestoredFromDatapath` is se to true for each backend when agent starts up (that's true). But while backends are restored only function `NewBackendWithState() is executed, nothing else. Inside that function the state is correctly recovered from the bpf maps as you know. But that's it and I am not changing anything there. 

The runtime tests passed just fine without any hiccup. 
```
Ran 67 of 414 Specs in 3751.755 seconds
SUCCESS! -- 67 Passed | 0 Failed | 9 Pending | 338 Skipped

Test Suite Passed
```
.. I ran `make test` from test directory.

oblazek replied:
and this function `updateBackendsCacheLocked` where I removed the code path (the one we are talking about) is never executed while restoration so there can't be any issue (related to my changes) while the backend is restored.

aditighag replied:
I think you are testing only in the standalone LB mode? Runtime test doesn't set up a k8s cluster either. Anyway I consolidated the discussion below.

oblazek committed:
refactor: backend weights handler

RestoredFromDatapath flag has been used to set state of a backend
to the saved state when backend was being restored.
The logic has been simplified and this flag is not needed anymore.

Backend weights API has been blocked when k8s is enabled as
there is currently no svc annotation where weights would be defined.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

brb commented on Documentation/cmdref/cilium_service_update.md:
````diff
@@ -11,6 +11,7 @@ cilium service update [flags]
 ### Options
 
 ```
+      --backend-weights uints       Backend weights (1 default, 0 means maintenance state) (default [])
````
We should mention that weights work only for Maglev.

oblazek replied:
sg

brb commented on pkg/service/service.go:
```diff
@@ -1369,17 +1370,23 @@ func (s *Service) restoreServicesLocked() error {
+					scopedLog.WithError(err).Warning("Unable to upsert into the Maglev BPF map.")
+					continue
+				}
+				if svc.NatPolicy == lb.SVCNatPolicyNat46 {
```
Why is this if-branch needed?

oblazek replied:
my bad during rebasing.. let me fix that

brb reviewed (commented):
Thanks, some minor comments.

aditighag commented on daemon/cmd/loadbalancer.go:
```diff
@@ -55,11 +55,14 @@ func (h *putServiceID) Handle(params PutServiceIDParams) middleware.Responder {
 		if err != nil {
 			return api.Error(PutServiceIDInvalidBackendCode, err)
 		}
+		if params.Config.WeightsUpdated {
```
Is this field required? Can't we derive it from `len(params.Config.WeightsUpdated)`?

oblazek replied:
removed, I believe I simplified the logic a bit :)

aditighag commented on cilium/cmd/service_update.go:
```diff
@@ -28,6 +28,7 @@ var (
 	frontend           string
 	backends           []string
 	backendStates      []string
+	backendWeights     []uint
```
Should this be defined as `uint16` to avoid type casting below?

oblazek replied:
it's not that simple as `spf13/pflags` - `UintSliceVarP` supports only `uint` :/.

aditighag commented on api/v1/openapi.yaml:
```diff
@@ -2613,6 +2626,9 @@ definitions:
           Update all services selecting the backends with their given states
           (id and frontend are ignored)
         type: boolean
+      weightsUpdated:
```
Is this required to be defined in the API yaml, or can it be derived from backend-weights?

oblazek replied:
removed

aditighag reviewed (commented):
Revisions look good, I've posted some minor comments.

aditighag commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -572,22 +570,20 @@ func NewBackend(id BackendID, protocol L4Type, ip net.IP, portNumber uint16) *Ba
-// NewBackendWithState creates the Backend struct instance from given params,
-// and sets the restore state for the Backend.
+// NewBackendWithState creates the Backend struct instance from given params.
 func NewBackendWithState(id BackendID, protocol L4Type, ip net.IP, portNumber uint16,
```
Why are we not setting the weight field to the default value here - `1`? This is called from the restore path. Do we assume that an external controller sets the weights after agent restart?

oblazek replied:
we should :+1:

aditighag reviewed (commented):
Changes look good. I've posted a question around setting the default weight for a backend. Should be good to merge the PR once the question is resolved.

oblazek committed:
fix: backend state related unit tests

When computing maglev algorithm and respecting weights it is necessary to pass the whole
Backend struct into UpsertService function (instead of previously passing only BackendIDs).
As backends are now passed by pointer it is necessary to reset backend states in corresponding
unit tests if state has changed.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

aditighag commented on pkg/service/service_test.go:
```diff
@@ -822,6 +822,11 @@ func (m *ManagerTestSuite) TestUpsertServiceWithTerminatingBackends(c *C) {
 		Name:                      lb.ServiceName{Name: "svc1", Namespace: "ns1"},
 	}
 
+	// Reset state as backends are pointers to lb.Backend
+	p.Backends[0].State = lb.BackendStateActive
+	p.Backends[1].State = lb.BackendStateActive
+	p.Backends[2].State = lb.BackendStateActive
+
```
Since backends are passed as pointers now, I would suggest making a deep copy of the test backend objects in unit tests, instead of resetting individual members of the global backend objects in every test case.
There are many instances of `DeepCopy` being called in tests files in the cilium repo in case you were looking for a reference.

Edit: I noticed that the change from `lb.Backend` to `*lb.Backend` was introduced in some other commit, so don't worry about making these changes. 

We should be ready to merge the PR once you rebase, and we re-run the l4lb test again, provided the test passes.
4.9 failure is same as - https://github.com/cilium/cilium/issues/20288. This is also fixed in latest master.

oblazek replied:
ok thanks, that make sense

aditighag commented on pkg/service/service_test.go:
```diff
@@ -928,8 +933,14 @@ func (m *ManagerTestSuite) TestRestoreServiceWithTerminatingBackends(c *C) {
 		Name:                      lb.ServiceName{Name: "svc1", Namespace: "ns1"},
 	}
 
+	// Reset state as backends are pointers to lb.Backend
+	p.Backends[0].State = lb.BackendStateActive
+	p.Backends[1].State = lb.BackendStateActive
+	p.Backends[2].State = lb.BackendStateActive
+
```
Same as above.

aditighag reviewed (approved):
I've left a comment in the test file. If the CI passes, no need to address it in this PR. But please address it in a follow-up PR.

squeed commented on api/v1/openapi.yaml:
```diff
@@ -2509,6 +2509,14 @@ definitions:
           Indicator if this backend is preferred in the context of clustermesh service affinity. The value is set based
           on related annotation of global service. Applicable for active state only.
         type: boolean
+  BackendWeight:
```
This seems a bit awkward -- is there a reason this is not a property on the BackendAddress object?

oblazek replied:
this was proposed by the initial author, but makes more sense to me to be part of `BackendAddress` object :+1:

oblazek replied:
let me fix that and the logic around that

oblazek replied:
removed this and made it part of BackendAdress object

squeed commented on api/v1/openapi.yaml:
```diff
@@ -2581,6 +2589,11 @@ definitions:
         type: array
         items:
           "$ref": "#/definitions/BackendAddress"
+      backend-weights:
```
I assume this needs to be the same length as `backend-addresses`? Can you expand this in the docs?

oblazek replied:
true -- either same length or nothing (which means default weight 1), yeah I can do that

oblazek replied:
removed, now part of `BackendAddress` object

oblazek committed:
fix: maglev table backend restoration

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
refactor: api handler

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

squeed commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -582,16 +580,15 @@ func NewBackend(id BackendID, protocol L4Type, ip net.IP, portNumber uint16) *Ba
+		ID:       id,
+		L3n4Addr: L3n4Addr{IP: ip, L4Addr: *lbport},
+		State:    state,
+		Weight:   1,
```
Another thought occurs to me: Should the default weight be higher? 1000? 32768? That way one can decrease the weight of a single backend relative to the pool.

Also, the default weight should be a constant (even if it's 1) throughout the code.

oblazek replied:
sounds good, I will make it a constant.. about the default value, we usually use weights between 0 and 200 and that is sufficient :) Does 100 make sense as a default value?

squeed replied:
100 sounds great, thanks.

oblazek replied:
done

oblazek committed:
feat: set default backend weight to 100 and make it a const

* Fix related unit tests.
* Make a pointer out of weight in BackendAddress object

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

squeed commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -573,7 +574,7 @@ func NewL3n4AddrFromModel(base *models.FrontendAddress) (*L3n4Addr, error) {
 func NewBackend(id BackendID, protocol L4Type, ip net.IP, portNumber uint16) *Backend {
 	lbport := NewL4Addr(protocol, portNumber)
 	b := Backend{
```
Should you add a DefaultWeight here too?

oblazek replied:
yeah you are right! thanks

oblazek replied:
done

squeed commented on api/v1/models/backend_address.go:
```diff
@@ -39,6 +39,9 @@ type BackendAddress struct {
 	State string `json:"state,omitempty"`
+
+	// Backend weight
+	Weight uint16 `json:"weight,omitempty"`
```
I just realized: doesn't this need to be a pointer, since 0 is a valid value?

oblazek replied:
you mean to distinguish between unset and 0 value?

squeed replied:
exactly :-)

oblazek replied:
ok sg, I will update the openapi scheme with `x-nullable: true`

oblazek committed:
fix: logic to allow backend weights moved to service_update.go file

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

squeed reviewed (approved):
Looks great! Thanks!
